core_v2/sources/v2_domains.move

```diff
@@ -0,0 +1,1315 @@
+    /// The caller is not the owner of the name, and is not authorized to perform the action
+    const ENOT_OWNER_OF_NAME: u64 = 9;
+    /// The domain name is too long- it exceeds the configured maximum number of utf8 glyphs
+    const ENAME_TOO_LONG: u64 = 10;
+    /// The domain is too short.
+    const ENAME_TOO_SHORT: u64 = 11;
+    /// The domain name contains invalid characters: it is not a valid domain name
+    const ENAME_HAS_INVALID_CHARACTERS: u64 = 12;
+    /// The required `register_domain_signature` for `register_domain` is missing
+    const EVALID_SIGNATURE_REQUIRED: u64 = 16;
+    /// The name is not expired in 6 months, thus not eligible for renewal
+    const EDOMAIN_NOT_AVAILABLE_TO_RENEW: u64 = 18;
+    /// The subdomain is not eligible for renewal
+    const ESUBDOMAIN_IS_AUTO_RENEW: u64 = 19;
+    /// The name is expired
+    const ENAME_EXPIRED: u64 = 20;
+    /// The subdomain not exist
+    const ESUBDOMAIN_NOT_EXIST: u64 = 21;
+    /// The name is not a subdomain
+    const ENOT_A_SUBDOMAIN: u64 = 22;
+    /// The subdomain expiration can be set any date before the domain expiration
+    const ESUBDOMAIN_EXPIRATION_PASS_DOMAIN_EXPIRATION: u64 = 24;
+    /// The duration must be whole years
+    const EDURATION_MUST_BE_WHOLE_YEARS: u64 = 25;
+    /// The subdomain expiration policy is included in the enum SUBDOMAIN_POLICY_*
+    const ESUBDOMAIN_EXPIRATION_POLICY_INVALID: u64 = 26;
+    /// Caller must be the router
+    const ENOT_ROUTER: u64 = 27;
+    /// Cannot register subdomain while its domain has expired
+    const ECANNOT_REGISTER_SUBDOMAIN_WHILE_DOMAIN_HAS_EXPIRED: u64 = 28;
+    /// Cannot transfer subdomain while its domain has expired
+    const ECANNOT_TRANSFER_SUBDOMAIN_WHILE_DOMAIN_HAS_EXPIRED: u64 = 29;
+    /// The domain is expired
+    const EDOMAIN_EXPIRED: u64 = 30;
+    /// Name is expired and out of grace period
+    const ECANNOT_RENEW_NAME_THAT_IS_EXPIRED_AND_PAST_GRACE_PERIOD: u64 = 31;
+
+    #[resource_group(scope = global)]
+    struct ObjectGroup { }
+
+    /// Tokens require a signer to create and we want to store global resources. We use object to achieve both
+    struct DomainObject has key {
+        extend_ref: ExtendRef,
+    }
+
+    #[resource_group_member(group = aptos_names_v2::v2_domains::ObjectGroup)]
+    struct NameRecord has key {
+        domain_name: String,
+        expiration_time_sec: u64,
+        target_address: Option<address>,
+        transfer_ref: TransferRef,
+        registration_time_sec: u64,
+        // Currently unused, but may be used in the future to extend with more metadata
+        extend_ref: ExtendRef,
+    }
+
+    #[resource_group_member(group = aptos_names_v2::v2_domains::ObjectGroup)]
+    /// This is a subdomain extension that is only used for subdomains
+    struct SubdomainExt has key {
+        subdomain_name: String,
+        subdomain_expiration_policy: u8,
+    }
+
+    struct ReverseRecord has key {
```

Comment:
We probably should have had this in its own resource group so we can upgrade it in the future, if we deem it necessary. Is it too late?